### PR TITLE
scripts : add hf.sh helper script

### PR DESCRIPTION
**Pull request by ggerganov**
Helper script to download an HF model file using `wget` or `curl` (whichever is available). Supports resuming downloads.

Sample usage:

```bash
./main \
  -m $(./scripts/hf.sh --repo TheBloke/Mixtral-8x7B-v0.1-GGUF --file mixtral-8x7b-v0.1.Q4_K_M.gguf) \
  -p "I believe the meaning of life is" -n 64

./main \
  -m $(./scripts/hf.sh https://huggingface.co/TheBloke/Mixtral-8x7B-v0.1-GGUF/blob/main/mixtral-8x7b-v0.1.Q4_K_M.gguf) \
  -p "I believe the meaning of life is" -n 64

./main \
  -m $(./scripts/hf.sh --url https://huggingface.co/TheBloke/Mixtral-8x7B-v0.1-GGUF/blob/main/mixtral-8x7b-v0.1.Q4_K_M.gguf) \
  -p "I believe the meaning of life is" -n 64
```

old attempt: #4735

### CODE REVIEW

File: scripts/hf.sh [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+is_url=false
+
+if [[ ${#url} -gt 22 ]]; then
+    if [[ ${url:0:22} == "https://huggingface.co" ]]; then
+        is_url=true
+    fi
+fi
+
+if [ "$is_url" = false ]; then
+    exit 0
```
Comment (Nindaleth) [inline scripts/hf.sh:70]: a log could be added here? otherwise non-HF/mistyped URIs will quit silently

**Review by Vaibhavs10 — COMMENTED** [review]
Love itt! Conceptually I wonder if it'd make more sense to add support for providing the repo name as well as the file name and we make the URL in the script.

This helps remove any possibility of people executing malicious code or accidentally putting in the wrong URL too.

`model-repo` & `quant-file` (or something like that).

`./main \
  -m $(./scripts/hf.sh --model-repo TheBloke/Mixtral-8x7B-v0.1-GGUF --quant-file mixtral-8x7b-v0.1.Q4_K_M.gguf) \
  -p "I believe the meaning of life is" -n 64`

What do you think?

**Comment by ggerganov** [comment]
Just added the optional `--repo` and `--file` args

I prefer `--url` because I can just copy paste the URL from the browser and don't have to split it into parts. But both options are available now.

> This helps remove any possibility of people executing malicious code or accidentally putting in the wrong URL too.

This script is mainly a helper for `llama.cpp` examples. It's not meant for production use, so I don't think we have to worry about wrong-doings here.

File: scripts/hf.sh [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+#   ./main -m $(./examples/hf.sh --repo TheBloke/Mixtral-8x7B-v0.1-GGUF --file mixtral-8x7b-v0.1.Q4_K_M.gguf)
+#
+
+# all logs go to stderr
+function log {
+    echo "$@" 1>&2
+}
+
+function usage {
+    log "Usage: $0 [[--url] <url>] [--repo <repo>] [--file <file>] [-h|--help]"
```
Comment (slaren) [inline scripts/hf.sh:17]: Is the extra `[` intentional?
```suggestion
    log "Usage: $0 [--url] <url>] [--repo <repo>] [--file <file>] [-h|--help]"
```

**Review by Vaibhavs10 — APPROVED** [review]
Brilliant thanks for iterating! This looks good to me! I think this could help the community quite a bit! ⚡

**Review by Nindaleth — APPROVED** [review]

File: scripts/hf.sh (hunk 2 above)
Comment (ggerganov) [reply scripts/hf.sh:17]: Yes, it's a nested optional. You can either do:

```
--url https://...
```

or just

```
https://...
```

**Pull request merged**